core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -217,7 +219,8 @@ class ReplicaManager(val config: KafkaConfig,
 
   def startup() {
     // start ISR expiration thread
-    scheduler.schedule("isr-expiration", maybeShrinkIsr, period = config.replicaLagTimeMaxMs, unit = TimeUnit.MILLISECONDS)
+    // A follower can lag behind leader for up to config.replicaLagTimeMaxMs x (1 + 20%) before it is removed from ISR
+    scheduler.schedule("isr-expiration", maybeShrinkIsr, period = config.replicaLagTimeMaxMs / 5, unit = TimeUnit.MILLISECONDS)
```

Comment:
Now that we remove the check in `maybeExpandIsr()`, it is probably more important to run maybeShrinkIsr() sooner than later. Note that if we don't shrink ISR soon enough, it is more likely to have message duplication. Here is the scenario:

- replication factor = 3 and min isr = 2. Current ISR set includes all 3 replicas.
- Replica max lag = 10 seconds, request timeout = 12 seconds, we run maybeShrink() every 5 seconds.
- At time t, A follower stops fetching but doesn't go offline.
- At time t, a produce request with ack = -1 is received by leader and the message is appended to log. Now leader will for all replicas in the ISR to fetch this message before replying to producer.
- At time t + 12, leader sends request timeout error to producer. producer retries. The message is appended to log again.
- At time t + 15, that follower is removed from ISR. HW is increased to include that message twice. Consumer will receive that message twice.

To prevent this scenario we need request timeout to be larger than max replica lag. If we allow maybeShrink() to run infrequently, then the producer's request timeout needs to be larger, which is not desirable.